File: 프로그래머스/lv2/42578. 의상/의상.js
// 핵심 로직: 각각의 종류의 개수+1들을 모두 곱하고, 마지막에 -1
  // 상의3개, 모자2개인 경우는 4*3-1이다. why?
  // 상의가 3개이지만, 상의에서 나올 수 있는 선택지는 4개이다. 바로, '상의선택x'경우도 +!
  // 단, 모-든 옷종류가 선택x를 택할 수는 없으므로 이 경우를 마지막에 -1해서 빼주는 것

function solution(clothes) {
  let obj = {}
  for (let cloth of clothes) {
    if (!obj[cloth[1]]) obj[cloth[1]] = 1
    else obj[cloth[1]] += 1
  };

  let values = Object.values(obj);
  let answer = 1;

  for (let value of values) {
    answer *= (value+1)
  };

  return answer-1
};


console.log(solution([["yellow_hat", "headgear"], ["blue_sunglasses", "eyewear"], ["green_turban", "headgear"]]));
console.log(solution([["crow_mask", "face"], ["blue_sunglasses", "face"], ["smoky_makeup", "face"]]));

File: 프로그래머스/2/42578. 의상/의상.py
# 모자가 2종류, 코트가 1종류라면?
    # (x1,a,b) / (x2,c) <- 이렇게 x1,x2를 한개씩 넣어줘야 한다
    # x의 의미: blank. 한 종류의 의상만 쓸 수도 있으므로
    # x1x2,x1c, ax2,ac, bx2,bc -> 여기서 x1x2는 빼야 함
    # 따라서 return (k1+1)*(k2+1)-1! 
# clothes를 한방에 dict로 바꾸는 메서드 없을까
from math import prod


def solution(clothes):
    dict = {}

    for clothe in clothes:
        if clothe[1] not in dict:
            dict[clothe[1]] = 1
        else:
            dict[clothe[1]] += 1
    
    value_list = list(dict.values())
    value_list = [x+1 for x in value_list]
    return prod(value_list)-1



print(solution([["yellow_hat", "headgear"], ["blue_sunglasses", "eyewear"], ["green_turban", "headgear"]]))
print(solution([["crow_mask", "face"], ["blue_sunglasses", "face"], ["smoky_makeup", "face"]]))

# 레퍼런스1 - 아. 너무 복잡하게 생각했다;;
    # 1)중간에 +1해서 빈종류 생각하기전에, 걍 애초에 시작부터 수를 2로 두면 됨
    # 2)곱하기 걍 반복문써도 되는데 끙;;
# def solution(clothes):
#     clothes_type = {}

#     for c, t in clothes:
#         if t not in clothes_type:
#             clothes_type[t] = 2
#         else:
#             clothes_type[t] += 1

#     cnt = 1
#     for num in clothes_type.values():
#         cnt *= num
#     return cnt - 1

# 레퍼런스2
# def solution(clothes):
#     from collections import Counter
#     from functools import reduce
#     cnt = Counter([kind for name, kind in clothes])
#     answer = reduce(lambda x, y: x*(y+1), cnt.values(), 1) - 1
    
#     return cnt